### PR TITLE
sql-sqlite: move every SQLite schema into numbered .sql migrations

### PR DESCRIPTION
**Pull request by mykola-vrmchk**
## What

Moves every SQLite schema in the repo out of TypeScript string literals into numbered, immutable `.sql` migration files, applied through a per-store history table by `@effect/sql`'s `Migrator`. **41 production DDL literals → 0**, across 18 stores in 9 packages.

New in `@dxos/sql-sqlite`:

- **`SqlMigrations`** — splits a multi-statement `.sql` script quote- and comment-aware, and applies it. Required because SQLite's `prepare` compiles exactly one statement and none of the three platform clients expose an `exec`.
- **`SqlTransaction.clientLayer`** — a `SqlClient` whose `withTransaction` delegates to the `SqlTransaction` service, so the migrator runs where literal `BEGIN`/`COMMIT` are forbidden.

Each store gets `src/migrations/{000N_*.sql, index.ts}`, a `*.sql?raw` ambient declaration, and a `migrate` that is one `Migrator.make({})` call.

Converted: `feed`, `index-core`, `echo-host`, `client-services`, `keyring`, `teleport-extension-object-sync`, `crawler`, `pipeline-rdf`, `pipeline-discord`.

## Why

Schema evolution was `CREATE TABLE IF NOT EXISTS` re-run on every open, plus `ALTER TABLE` wrapped in `catchAll` to swallow "duplicate column". That works, but it is write-only: nothing recorded what had been applied, no single artifact described a database's shape, and `objectMeta` had already drifted into a *family* of shapes depending on when a profile was created.

## Decisions a reviewer should know

**No Prisma, and the blocker is the browser.** Prisma Client reaches a database through a driver adapter; none exists for wa-sqlite over OPFS (the main client database) or for Durable Object `ctx.storage.sql` — D1's adapter does not cover DO storage. These packages must work in node, browser and workerd from one codebase. Prisma was built and piloted here, then removed; as a schema-authoring tool alone it earned too little to keep. Full reasoning and the adapter table: [`.agents/projects/sql-migrations/DESIGN.md`](https://github.com/dxos/dxos/blob/claude/sql-types-string-literals-audit-c269d2/.agents/projects/sql-migrations/DESIGN.md).

**Existing databases need no baselining.** Every `CREATE` in each `0001_init.sql` carries `IF NOT EXISTS`, so a database from an earlier release applies migration 1 as a recorded no-op with rows preserved. The clause is load-bearing and cannot be caught by any schema-equivalence check — a test asserts it after splitting each file.

**`index-core` needed a code migration.** `objectMeta`'s four historical `ADD COLUMN`s are not expressible as plain SQL that is correct for every vintage (SQLite has no `ADD COLUMN IF NOT EXISTS`, and an unconditional ALTER fails on fresh databases too). `0002_missing_columns` probes `PRAGMA table_info` and adds exactly what is missing; `0003_indexes` follows because it references those columns.

**`Migrator.make({})`, not the platform `SqliteMigrator`** — those additionally require `FileSystem`/`Path`/`CommandExecutor` and cannot run in the browser.

## Testing

Per package: the `IF NOT EXISTS` invariant, manifest coverage, a legacy database applying as a recorded no-op, and second-run-is-a-no-op — with expected ids derived from the manifest rather than hard-coded. `index-core` adds vintage tests (fresh / oldest / partially-migrated) converging on the code-derived shape with rows preserved.

Beyond the suite, three independent checks:

1. **Replay audit** — the exact DDL `origin/main` ships, replayed for all 18 stores, then this branch's migrations applied on top: `PRAGMA table_info`/`index_list`/`foreign_key_list` match a fresh database.
2. **Cross-version interoperability** — for all 9 packages, old code wrote a file database through its real API, then this branch's code migrated it in place, read every row back, and wrote new rows on top.
3. **Durable Objects** — dxos CI cannot cover this path (no `ts-test-workerd` in `feed` or `sql-sqlite`), so it is verified in [edge#790](https://github.com/dxos/edge/pull/790) with linked packages: `feed-migrations.workerd.test.ts` for transaction control and `index-core-migrations.workerd.test.ts` for the `PRAGMA` probe, which is the only `PRAGMA` dxos runs through `@effect/sql-sqlite-do`. db-service 142/142.

### CODE REVIEW

File: .moon/tasks/tag-prisma.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+tasks:
+  prisma:
+    options:
+      internal: true
+    command: node $workspaceRoot/scripts/prisma-generate-sql.mjs
+    inputs:
+      - /scripts/prisma-generate-sql.mjs
+      - 'prisma/**/*.prisma'
+    outputs:
+      - src/migrations/
```
[the remark this answers is not in the document]
Comment (mykola-vrmchk) [reply .moon/tasks/tag-prisma.yml]: Fixed in e9b0f71268 — real bug, thanks. Narrowed to `src/migrations/*.sql`. Verified the glob resolves (`outputGlobs: packages/core/echo/feed/src/migrations/*.sql`) and functionally: deleted both `.sql` files, ran `build`, moon restored them from cache and left `index.ts` untouched.

File: packages/core/echo/feed/src/feed-store.ts [hunk 2]
```diff
@@ -61,58 +62,17 @@ export class FeedStore {
...
-          lastPulledPosition INTEGER NOT NULL DEFAULT -1,
-          PRIMARY KEY (spaceId, feedNamespace)
-      )`;
-    }).pipe(Effect.withSpan('FeedStore.migrate')),
+    SqlMigrator.run({
+      table: MIGRATIONS_TABLE,
+      migrations: MIGRATIONS,
+      baseline: { throughId: 1, when: SqlMigrator.tableExists('feeds') },
+    }).pipe(Effect.asVoid, Effect.withSpan('FeedStore.migrate')),
   );
```
[the remark this answers is not in the document]
Comment (mykola-vrmchk) [reply packages/core/echo/feed/src/feed-store.ts:85]: Fixed in e9b0f71268. Correct that `Effect.asVoid` does not narrow the failure channel. Rather than widening the documented type, `FeedStore.migrate` now dies on `MigrationError` — a malformed bundled manifest is a defect, not something a caller can recover from — so the signature stays `SqlError`-only and existing callers are unaffected. `schema-harness.ts` mirrors it so the tests exercise the same handling.

File: packages/common/sql-sqlite/src/SqlMigrations.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+ * DDL is spliced in with `sql.literal`, since an interpolated string would otherwise be sent as
+ * a bound parameter rather than as SQL.
+ *
+ * @example
+ * ```typescript
+ * import schema from './migrations/schema.sql?raw';
+ *
+ * yield* SqlMigrations.apply(schema);
+ * ```
+ */
```
[the remark this answers is not in the document]
Comment (mykola-vrmchk) [reply packages/common/sql-sqlite/src/SqlMigrations.ts:49]: Fixed in e9b0f71268. Both were stale from the pilot's first pass, which had no version table. `SqlMigrations.apply` is now documented as the raw executor that tracks nothing, pointing to `SqlMigrator.run` for a store's schema, and the `@example` uses `0001_init.sql`. The moon task comment now says `SqlMigrator.run`.

File: packages/core/echo/feed/prisma/schema.prisma [hunk 4]
```diff
@@ -0,0 +1,69 @@
+// Schema for the feed store's tables. Generates `src/migrations/schema.sql` via
+// `moon run feed:prisma`; no prisma client is generated and none ships to the runtime.
```
[the remark this answers is not in the document]
Comment (mykola-vrmchk) [reply packages/core/echo/feed/prisma/schema.prisma]: Fixed in e9b0f71268. The header now says the task regenerates `snapshot.sql` and writes `0001_init.sql` only when no migration exists yet, and that existing migrations are immutable.

File: agents/superpowers/specs/2026-07-31-prisma-sql-ddl-design.md [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+
+Statement splitting is the only genuinely new logic and has its own unit tests.
+
+Verified in the built bundle: the `?raw` import is inlined as `var schema_default = "--\n…"`
+with no filesystem access, so the browser path holds.
+
+### Type mapping
+
+Verified against the real DDL by the spike — the emitted SQL is character-identical to the
+existing hand-written definitions:
```
[the remark this answers is not in the document]
Comment (mykola-vrmchk) [reply agents/superpowers/specs/2026-07-31-prisma-sql-ddl-design.md]: Fixed in e9b0f71268. This contradicted the same document's own list of differences. Replaced with the specific equivalence the tests assert: column name, type, nullability, PK position, default, index name/uniqueness/columns, and FK target and actions, read from `PRAGMA table_info` / `index_list` / `foreign_key_list`, with PK nullability excluded for the documented reason.

File: agents/superpowers/specs/2026-07-31-prisma-sql-ddl-design.md [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+- `crawl_run`'s `id INTEGER PRIMARY KEY CHECK (id = 1)` singleton constraint.
+
+### Drift
+
+The four `ALTER TABLE objectMeta ADD COLUMN` statements (`parent`, `createdAt`, `updatedAt`,
+`queueNamespace`) **must be retained** as numbered migrations. Because migration 1 is
+`CREATE TABLE IF NOT EXISTS` and is stamped onto an existing profile rather than run, that
+profile keeps whatever columns it was created with. Dropping the ALTERs would silently strip
+those four columns from older profiles — the single highest-risk item in the change, since it
+affects real local-first user data in OPFS and fails silently.
```
[the remark this answers is not in the document]
Comment (mykola-vrmchk) [reply agents/superpowers/specs/2026-07-31-prisma-sql-ddl-design.md]: Fixed in e9b0f71268 — the wording was wrong, not just imprecise. `CREATE TABLE IF NOT EXISTS` cannot alter an existing table, so dropping the ALTERs would leave those columns missing rather than remove anything. Reworded.

File: agents/superpowers/specs/2026-07-31-prisma-sql-ddl-design.md [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+
+Prisma 7 removes `url` from the datasource block and requires a `prisma.config.ts`. Adopting
+7 here would diverge from edge for no benefit. Confirmed by spike: 6.19.3 works with the
+schema shape below; 7.9.1 rejects it.
+
+## Architecture
+
+Per-package layout:
+
+```
```
[the remark this answers is not in the document]
Comment (mykola-vrmchk) [reply agents/superpowers/specs/2026-07-31-prisma-sql-ddl-design.md]: Fixed in e9b0f71268 — fence tagged `text`.

File: .changeset/quiet-donkeys-repeat.md [hunk 8]
```diff
@@ -0,0 +1,5 @@
+---
+'@dxos/sql-sqlite': minor
+---
+
+Add `SqlMigrations` and `SqlMigrator` for applying versioned SQL migrations recorded in a per-store history table, including baselining for databases created before migration tracking existed. The feed store's schema is now authored as a Prisma schema and applied as numbered migrations rather than inline DDL.
```
[the remark this answers is not in the document]
Comment (mykola-vrmchk) [reply .changeset/quiet-donkeys-repeat.md]: Declining — I think this one is incorrect. `@dxos/sql-sqlite` *is* a package changed by this PR, and the primary one: it gains `SqlMigrations` and `SqlMigrator` plus two new export subpaths, which is the new public API here. It is publishable (`publishConfig.access: public`, not private) and in Group A. Per `agents/instructions/changesets.md`, the rule is to name **one** package you changed, and naming any Group A member bumps the whole group — so `@dxos/feed` is already covered and adding it would be the enumeration the guide warns against. Happy to switch if the intent is that the changelog entry should read from `feed`'s perspective, but the bump is identical either way.

File: agents/superpowers/specs/2026-07-31-prisma-sql-ddl-design.md [hunk 9]
```diff
@@ -0,0 +1,428 @@
...
+succeed.
+
+`SqlMigrator.run` therefore takes an optional `baseline: { throughId, when }`. When the history is
+empty and `when` holds — typically `tableExists('<a table the store owns>')` — migrations up to
+`throughId` are **recorded without being executed**. This is the equivalent of `prisma migrate
+resolve --applied`, which is unavailable to us because it needs a live database and an engine.
+
+Correctness here cannot be observed from the resulting schema, since a stamped and a re-run
+migration leave the same tables. The observable difference is `run`'s return value: the list of
+migrations actually executed, which excludes anything stamped.
```
[the remark this answers is not in the document]
Comment (mykola-vrmchk) [reply agents/superpowers/specs/2026-07-31-prisma-sql-ddl-design.md]: Agreed, and already the stated position — see the Drift section: `feed` baselines at `throughId: 1` only because its schema never changed, and the spec explicitly says a fixed `throughId` cannot be right for every database and that `index-core` will need to probe live columns or keep the ALTER migrations individually tolerant. I have not generalised `feed`'s predicate. Left as a decision for the `index-core` conversion rather than pre-committing to a mechanism here.

File: agents/superpowers/specs/2026-07-31-prisma-sql-ddl-design.md [hunk 10]
```diff
@@ -0,0 +1,428 @@
...
+DDL and never point Prisma at a live database. Verified safe: no foreign key crosses a
+package boundary — the only FK, `blocks` → `feeds`, is internal to `feed`.
+
+### Migrations are numbered, immutable, and tracked in a migrations table
+
+Each store keeps its own history table (`feed_migrations`, …), since several packages share one
+physical database. Migrations are numbered `.sql` files, listed explicitly in an `index.ts`
+manifest so ids and ordering are reviewable in a diff. Once written they are never edited: they
+are recorded as applied and never re-run, so changing one would diverge from databases that
+already ran it.
```
[the remark this answers is not in the document]
Comment (mykola-vrmchk) [reply agents/superpowers/specs/2026-07-31-prisma-sql-ddl-design.md]: Valid gap, now recorded rather than implemented. e9b0f71268 adds it to the Testing section as a required CI check: reject duplicate or non-increasing ids, and reject diffs that modify an already-committed migration. Both are undetectable at runtime — the table stores no checksum and advances a high-water mark — and, as you note, the snapshot test does not catch an edited applied migration, since the edit and the schema still agree. Until that check exists both rules are convention only, which the spec now says explicitly. Building it belongs with the remaining 8 packages, where there will be more than one manifest to check.

File: scripts/prisma-generate-sql.mjs [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+  .filter((entry) => /^\d+_.*\.sql$/.test(entry))
+  .sort();
+
+if (migrations.length === 0) {
+  writeFileSync(INITIAL_PATH, HEADER + schema);
+  console.log(`Wrote ${relative(process.cwd(), INITIAL_PATH)}`);
+  process.exit(0);
+}
+
+const replayed = describe(migrations.map((entry) => readFileSync(join(MIGRATIONS_DIR, entry), 'utf8')).join('\n'));
```
[the remark this answers is not in the document]
Comment (mykola-vrmchk) [reply scripts/prisma-generate-sql.mjs]: Addressed in de8c53a658, though not where you pointed — `scripts/prisma-generate-sql.mjs` has since been deleted, so there is no build-time replay left to fix. The underlying hole was real and survived that deletion: a `.sql` file missing from the `MIGRATIONS` manifest never runs, silently. There is now a test asserting the manifest lists every `.sql` on disk. The manifest stays explicit rather than globbed at runtime so a bundler cannot change what ships.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: agents/superpowers/specs/2026-07-31-prisma-sql-ddl-design.md [hunk 12]
```diff
@@ -0,0 +1,476 @@
...
+`prebuild` meta-task (edge does), so the task is attached by appending to `build`'s `deps` —
+moon merges inherited `deps` by append, which was verified: `feed:build` lists `feed:prisma`
+alongside the deps declared in `tag-ts-vite-build.yml`.
+
+`scripts/prisma-generate-sql.mjs` runs `prisma migrate diff --from-empty --to-schema-datamodel` and
+writes `0001_init.sql` only when no migration exists yet — it never overwrites a migration. On every
+other run it checks the initial migration still carries `IF NOT EXISTS` and performs the replay
+comparison above. It sets
+`PRISMA_HIDE_UPDATE_MESSAGE=1`, since the CLI otherwise prints an upgrade banner on every
+invocation, and aborts if prisma emits no DDL, so a prisma failure cannot empty committed SQL.
```
[the remark this answers is not in the document]
Comment (mykola-vrmchk) [reply agents/superpowers/specs/2026-07-31-prisma-sql-ddl-design.md]: Addressed in de8c53a658. The generator is gone, so there is no first-generation path to guard — but you were right that the invariant was unenforced. `IF NOT EXISTS` on every CREATE in the initial migration is now asserted by a test, which is the only place it can be checked: it is added by hand and does not change the resulting schema, so the equivalence assertions are blind to it. Verified by mutation — stripping the clauses fails it, naming all 8 statements.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: scripts/prisma-generate-sql.mjs [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+}
+
+// The initial migration is the one that can be stamped onto a database that already holds these
+// tables, so every CREATE in it must tolerate that. Verified rather than rewritten: the file stays
+// exactly what a reviewer read, and regenerating it (which drops the clause, since prisma emits
+// bare statements) fails here instead of silently losing the guarantee.
+const initial = readFileSync(join(MIGRATIONS_DIR, migrations[0]), 'utf8');
+const bare = initial
+  .replace(/--[^\n]*/g, '')
+  .match(/\bCREATE\s+(?:VIRTUAL\s+TABLE|UNIQUE\s+INDEX|TABLE|INDEX)\s+(?!IF\s+NOT\s+EXISTS)\S+/gi);
```
[the remark this answers is not in the document]
Comment (mykola-vrmchk) [reply scripts/prisma-generate-sql.mjs]: Good catch, and taken even though the script it refers to is deleted. The replacement check does exactly what you suggested: it runs `SqlMigrations.splitStatements` over the migration before inspecting statements, so a `;` or a CREATE keyword inside a comment or string literal cannot fool it. That was the specific fragility in the old comment-stripping regex.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: scripts/prisma-generate-sql.mjs [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+            .map((column) => `${column.name} ${column.type}${column.notnull ? ' NOT NULL' : ''}`),
+          indexes: db
+            .prepare(`PRAGMA index_list("${name}")`)
+            .all()
+            .filter((index) => !index.name.startsWith('sqlite_'))
+            .map((index) => `${index.name}${index.unique ? ' UNIQUE' : ''}`)
+            .sort(),
+        },
+      ]),
+    );
```
[the remark this answers is not in the document]
Comment (mykola-vrmchk) [reply scripts/prisma-generate-sql.mjs]: Not fixed — `scripts/prisma-generate-sql.mjs` and its `describe` comparison were deleted in 42576e0558/28db10099e, so the incomplete signature no longer exists. Worth noting the finding was correct: it omitted defaults, primary keys, foreign keys and index columns. The equivalent comparison that remains lives in `schema.test.ts` and does include all four.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: scripts/prisma-generate-sql.mjs [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+const HEADER = `--
+-- Generated from prisma/schema.prisma by scripts/prisma-generate-sql.mjs, then maintained by hand.
+--
+-- Add \`IF NOT EXISTS\` to every CREATE below. Prisma emits bare statements, which fail against a
+-- database that already holds these tables — reachable when a store's baseline predicate does not
+-- fire, e.g. a database left partly initialised by an earlier release.
+--
+-- Immutable from then on: it is recorded in the migrations table, never re-run, and its checksum is
+-- verified on every open. Add a new numbered migration instead of editing this one.
+--
```
[the remark this answers is not in the document]
Comment (mykola-vrmchk) [reply scripts/prisma-generate-sql.mjs]: Not fixed — the script this refers to no longer exists. Baselining is also gone: every statement in the initial migration is `IF NOT EXISTS`, so a database created before migration tracking simply applies it as a no-op, which removed the need for a baseline predicate entirely.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by mykola-vrmchk** [comment]
## Migration-safety audit (head `64407e0601`)

Two questions this PR must answer before merging, both verified empirically rather than by inspection.

### 1. Do existing databases survive these migrations?

**Yes — proven per store.** For each of the 16 converted stores: a database was built with the **old inline DDL extracted from `origin/main`** (including replaying the old `Effect.catchAll` tolerance for the `objectMeta` ALTERs), a probe row inserted into **every** table, then this PR's migration chain applied.

Asserted for every store: the chain applies without error; **no rows lost**; the resulting schema is identical to a fresh database's (columns, types, `NOT NULL`, PKs, indexes); and the chain is **re-appliable** (idempotent) on top of itself — which is exactly the existing-database situation, since these stores never had a history table.

Also covered: `objectMeta`'s **oldest vintage** (none of `parent`/`createdAt`/`updatedAt`/`queueNamespace`) — the `0002` code migration adds exactly the missing columns and converges. `index-core`'s test suite additionally pins the partially-migrated vintage with data in the pre-existing column.

Two known, benign divergences, both deliberate and pinned by tests:
- `feed`: fresh databases get `NOT NULL` on two `TEXT` primary keys where legacy tables allow NULL (existing tables keep their shape — `CREATE IF NOT EXISTS` never alters).
- ALTER-appended columns land in a different physical position than fresh `CREATE`s — true of the old code too, and invisible to name-based access.

### 2. Does it work in edge (Durable Objects)?

**Yes — verified against a linked build of this exact commit.** Every edge package consuming a converted store:

| suite | result |
|---|---|
| `db-service` (FeedSpace DO, indexer, automerge, subduction) | **140/140** |
| `hub-protocol` | 28/28 |
| `edge-feed` | 11/11 |
| `router` | 8/8 |
| `edge:test` query integration (retried in isolation) | 40/40 |

`db-service` includes a workerd suite asserting the `feed_migrations` history table in real DO SQL storage (edge #790). The two `edge:test` timeouts seen in one batch run are the known indexer flakes: they change identity between runs, the failing file passes in isolation against the same build, and edge `main`'s own CI is currently red with a different environment-dependent test.

Caveat that remains true: dxos CI has no workerd coverage for these packages, so the DO guarantees above rest on the edge-side verification until `ts-test-workerd` lands for `sql-sqlite` (recorded follow-up in the design doc).

🤖 Generated with [Claude Code](https://claude.com/claude-code)
